Raft Coalesced Heartbeats

tbg opened the pull request:
currently work in progress, see #229 for early discussions (further discussions should be held here).
- [x] basic prototype
- [x] keep group memberships on a per-node basis
- [x] receive heartbeats only from (supposed leader)
- [x] analogous mechanism for heartbeat responses
- [x] basic test setup
- [x] test w/ invalid heartbeats (no leader)
- [x] test w/ multiple partially-overlapping consensus groups
- [ ] consider correctness implications/corner cases (discussion below)

closes #229

cc @cockroachdb/developers

tbg committed:
coalesced heartbeats

* outgoing heartbeats originating from raft are ignored
* instead, each node will periodically send coalesced heartbeats
  to each of the other nodes.
* upon receipt of a heartbeat (which is coalesced, according to the above),
  individual heartbeats are fed into raft for each group that the sending
  node belongs to and follows a leader from the origin node.
* this necessitated adding a list of groupIDs to the internal `node` type.
* heartbeat responses: Raft's (potentially many) generated responses
  are bundled up and only a single response sent.
* fan out heartbeat responses to all groups considering themselves leader of
  a group of the sending node

tschottdorf commented:
More q's based on [the discussion](https://github.com/tschottdorf/cockroach/commit/666ff68cae6c57b2c893f26a4461c73f874533d0):

- we could theoretically have a situation in which the leader of a certain raft group deadlocks their goroutine (or something like that) while coalesced heartbeats keep going happily. Is that something we need to watch out for or should we simply assume that as long as multiraft looks ok, all raft groups are going strong? A coalesced heartbeat trigger could also check that groups we think are leading on our node actually send heartbeats (which we then discard).

- @bdarnell said:
    >Good point; we need to track leader status so we don't interfere with followers' candidate timers. I'm not as concerned about the cost of sending heartbeats when we're not the leader, but we should avoid fanning-out heartbeat responses when we're not the leader.`

  I'm interpreting this as "heartbeats are fanned out only to groups which believe that the sending node is the leader", which is what I've implemented in the next commit (The leader ID was already tracked in the group, I just needed to use it).

  Are there any subtle things that can go wrong still? Since now group members get heartbeats based on who they think their boss is, and not based on who actually is. For instance, once a node becomes a candidate it doesn't think of anyone as their leader, and consequently won't receive any heartbeats. Hence, one code path that leads to a Candidate->Follower transition is completely disabled, and while I think that one's not critical (might just cost some time), I wonder what else is in the bushes.

tschottdorf commented on multiraft/multiraft.go:
```diff
@@ -531,8 +625,16 @@ func (s *state) handleWriteResponse(response *writeResponse, readyGroups map[uin
+					// TODO(bdarnell): support removing nodes; fix double-application of initial entries
+					continue
+				case raftpb.ConfChangeUpdateNode:
+					// TODO(Tobias): What can be updated in such a request?
```
@bdarnell looked for this in the raft paper and in the protobufs but no dice. Is there an explanation of what this does anywhere?

bdarnell replied:
I'm not sure. This may be for updating the cc.Context field (which we don't use). I don't think we need to support this.

tschottdorf commented on storage/store.go:
```diff
@@ -229,7 +229,8 @@ type Store struct {
-	mu          sync.RWMutex     // Protects variables below...
+	mu sync.RWMutex // Protects variables below...
+	// TODO(Tobias) change int64 to uint64?
 	ranges      map[int64]*Range // Map of ranges by Raft ID
```
any particular reason why this is an int64 here?

spencerkimball replied:
That's what it started out as (used to be a 64 bit range ID). If we want to change Raft ID to a uint64 to match underlying raft implementation, that's fine, but it's definitely a separate change, as it threads through protos and a large amount of code.

bdarnell commented on multiraft/multiraft.go:
```diff
@@ -341,6 +354,42 @@ func (s *state) start() {
 			return
 
+		case req := <-s.reqChan:
+			_, groupOK := s.groups[req.GroupID]
```
Pull this out into a separate method instead of doing so much directly in the select/case.

bdarnell commented on multiraft/multiraft.go:
```diff
@@ -341,6 +354,42 @@ func (s *state) start() {
+			if req.Message.Type == raftpb.MsgHeartbeat {
+				// A heartbeat message is expanded into a heartbeat for each group
+				// that the remote node is a part of.
+				originNode, _ := s.nodes[req.Message.From]
```
You don't need `, _` for a map lookup - just omit the second return value if you don't need it.

bdarnell commented on multiraft/multiraft.go:
```diff
@@ -341,6 +354,42 @@ func (s *state) start() {
+				for groupID := range originNode.groupIDs {
+					// If the group is invalid or we don't think that the
+					// sending node is leading that group, don't propagate.
+					if g, ok := s.groups[groupID]; !ok || g.leader != req.Message.From || g.leader == s.nodeID {
```
The group should never be invalid here since originNode.groupIDs is private to multiraft.

tschottdorf replied:
done

bdarnell commented on multiraft/multiraft.go:
```diff
@@ -410,20 +493,31 @@ func (s *state) createGroup(op *createGroupOp) {
-		s.addNode(nodeID)
-	}
 
+	// TODO(Tobias): ok to create first the group, then the node?
```
Yes, I think so.

bdarnell commented:
We need similar coalescing/fan-out for MsgHeartbeatResp as well. This will be similar on the receiving sides (except that we fan out heartbeat responses to the groups that the receiving node is leader of instead of those that the sending node leads). For sending, we should just keep track of which nodes we have sent MsgHeartbeatResp and not send to the same node twice (per call to handleRaftReady).

tschottdorf commented:
In the works.

For testing, I would replace the local transport's client by one that supports intercepting (and holding back) outgoing messages, and probably add a demux for consuming it during tests (akin to the eventsDemux). With that we should be able to get pretty good testability.
Good idea? Am I overcomplicating things?

bdarnell commented:
Yeah, we should test this with a custom transport (see also blockingStorage in storage_test.go). I'm not sure whether the eventsDemux pattern makes sense here but if it turns out to be useful in your tests go for it.

tschottdorf commented:
Ok. I'll continue development in this PR, as I wouldn't want to merge anything before I can test it anyways.
Last commit contains my take at the heartbeat responses plus their fanout.

tbg committed:
add interceptable client and use in local test setup

tbg committed:
coalesced heartbeats test

* elect a leader and have it tick a number of times
* count and compare the heartbeats sent and responses received

TODO: add test that has followers tick as well; a closer interplay
between ticks and acknowledged messages would be needed to prevent
followers from turning into candidates.

tschottdorf commented:
@bdarnell Can you take another look?

tschottdorf commented on multiraft/multiraft.go:
```diff
@@ -180,6 +181,55 @@ func (m *MultiRaft) sendEvent(event interface{}) {
 
+// fanoutHeartbeat sends the given heartbeat to all groups which believe that
+// their leader resides on the sending node.
+func (s *state) fanoutHeartbeat(req *RaftMessageRequest) {
```
I looked into combining this with fanoutHeartbeatRequest, but in the end I liked it better as is.

bdarnell commented on multiraft/transport.go:
```diff
@@ -100,6 +103,18 @@ func NewLocalRPCTransport() Transport {
+// *interceptMessage to that channel and block until the message is
+// acknowledged. EnableEvents should be called before any clients are created,
+// i.e. prior to calls to Connect.
+func (lt *localRPCTransport) EnableEvents() {
```
I don't think we should build this into localRPCTransport. There should be a separate interceptingTransport type that either wraps an existing Transport or simply is a transport itself (I think the latter is probably better - you don't need the RPC machinery, just have Go call the ServerInterface method directly). Then this can all go in transport_test.go unless and until we have a use for it outside this package.

tschottdorf replied:
Ok, will go for the thin version without RPC then. Can you explain what bothers you about the current modifications? I thought about creating an extra client but then it just seemed silly to write the same thing twice with a feature that the other one might also want to use at times.

bdarnell replied:
My concern is mainly with the dramatically different implementation (including a change to blocking behavior) depending on whether EnableEvents has been called. Since EnableEvents can only be called immediately after creation it's effectively a different type anyway.

bdarnell commented on multiraft/multiraft_test.go:
```diff
@@ -28,23 +28,25 @@ import (
 func newTestCluster(size int, t *testing.T) *testCluster {
 	transport := NewLocalRPCTransport()
 	cluster := &testCluster{t: t}
+	cluster.transport = transport.(*localRPCTransport)
```
newTestCluster should take a Transport (possibly nil) as input; then you wouldn't need a separate start() method. We could also move to the pattern used in storage/range_test.go before the number of inputs gets too unwieldy.

bdarnell commented on multiraft/heartbeat_test.go:
```diff
@@ -0,0 +1,152 @@
+	leaderTickCount := 18
+	// Since waitForElection() currently triggers two ticks, it also triggers
+	// some heartbeats which remain unanswered (the sender isn't leader yet).
+	unansweredHeartbeats := (nodeCount - 1) * 2 // 2 because of 2 election ticks.
```
This is subtle. Can we not just reset the count of heartbeats after waitForElection returns?

tschottdorf replied:
I could probably have an extra consumer goroutine that processes messages until it receives the second heartbeat, at which it will make room for the goroutine that's in the test now.
But that would mean adapting even more to the fact that currently we have to Tick() to get an election, and it would test less (the current setup tests that non-leader heartbeats actually get discarded, albeit it's somehow muddled up in the actual test)

tschottdorf replied:
as mentioned in a later comment, I think I will do just that and separate the ticks-as-non-leader part.

bdarnell commented on multiraft/heartbeat_test.go:
```diff
@@ -0,0 +1,152 @@
+		for {
+			depleted := false
+			select {
+			case ev := <-cluster.events[i].LeaderElection:
```
This could use `ev, ok := <-cluster.events[i].LeaderElection` since I don't think you really need to wait for a millisecond timeout.

tschottdorf replied:
Changed. I had to modify eventDemux.stop() to also close its demuxed channels.

bdarnell commented on multiraft/heartbeat_test.go:
```diff
@@ -0,0 +1,152 @@
+	cluster.transport.EnableEvents()
+
+	cluster.start()
+	leaderTickCount := 18
```
Is the number 18 arbitrary or was it derived from something?

tschottdorf replied:
arbitrary. I had 10 in there while writing the test, then changed it around a bit. Probably will just run those sorts of tests with 3-4 different values in the final version.

bdarnell commented on multiraft/heartbeat_test.go:
```diff
@@ -0,0 +1,152 @@
+				cntR, heartbeatResponseCount)
+		}
+	}()
+	for iMsg := range ch {
```
I prefer to only use "range ch" when we expect the loop to terminate by the closing of the channel. It doesn't look like that's the case here, so I would use `for { iMsg := <-ch }`. This makes it more clear that the return at line 83 is the only exit (which means we could make it a break instead of a return and move the defer to the main body of the function).

tschottdorf replied:
good point, done.

bdarnell commented on multiraft/heartbeat_test.go:
```diff
@@ -0,0 +1,152 @@
+			t.Errorf("unexpected total number of heartbeat responses: %d (expected %d)",
+				cntR, heartbeatResponseCount)
+		}
+	}()
```
We should probably check that every element of outInDelta is zero at the end of the function.

tschottdorf replied:
that was the original intent, but now that we have those two extra ticks at the beginning... probably better to chew those extra ticks away as you suggested earlier and have something more reasonable here.

tschottdorf replied:
done.

bdarnell commented on multiraft/heartbeat_test.go:
```diff
@@ -0,0 +1,152 @@
+	cluster.createGroup(groupID, 3)
+	election := cluster.waitForElection(leaderIndex)
+
+	for i := 0; i < leaderTickCount; i++ {
```
Minor: I'd prefer to move the ticker to a separate goroutine and let the more complicated validation logic (including in particular the assertions) live on the main thread. This makes it more clear that there are no code paths that result in the assertions silently not getting executed.

tschottdorf replied:
that's currently awkward because we intercept everything (where really we only want to block and acknowledge heartbeats) but I will keep it in mind for the interceptingLocaTransport.

bdarnell replied:
Ah, I see. Nevermind, then.

bdarnell commented on multiraft/heartbeat_test.go:
```diff
@@ -0,0 +1,152 @@
+
+	groupID := uint64(1)
+	leaderIndex := 0
+	cluster.createGroup(groupID, 3)
```
With only one group, this test would pass even without any of the fanout changes, right? I think we should test with at least three groups, two of which have the same leader.

tschottdorf replied:
yes, will do that next after addressing the transport.

bdarnell commented on multiraft/multiraft.go:
```diff
@@ -180,6 +181,55 @@ func (m *MultiRaft) sendEvent(event interface{}) {
+			continue
+		}
+		if err := s.multiNode.Step(context.Background(), groupID, req.Message); err != nil {
+			log.V(4).Infof("node %v: coalesced heartbeat step failed for message %s", s.nodeID, groupID,
```
If we actually get an error return that should be a lot noisier than a V(4). I'd probably make it a log.Errorf() unless we start seeing spurious errors (I don't think this method ever actually returns a non-nil error unless you call it on a closed node)

tschottdorf replied:
I agree. On a related note, are there any guidelines with respect to the V(.) levels that we're currently following? I haven't seen them around much in the rest of the codebase, but it sure would be good to have them.

bdarnell replied:
There aren't really any guidelines; it's pretty ad-hoc for now. One rule of thumb is that it should be reasonable to enable V(1) across all modules at the same time without getting too overwhelmed, whereas any higher-V logging should probably be turned on for one file at a time with --vmodule.

tschottdorf commented:
@bdarnell did you see this above:

```
Are there any subtle things that can go wrong still? Since now group members get heartbeats based on who they think their boss is, and not based on who actually is. For instance, once a node becomes a candidate it doesn't think of anyone as their leader, and consequently won't receive any heartbeats. Hence, one code path that leads to a Candidate->Follower transition is completely disabled, and while I think that one's not critical (might just cost some time), I wonder what else is in the bushes.
```

tbg committed:
address @bdarnell's feedback

bdarnell commented:
I had missed that note the first time it appeared. It's good to be concerned about this; I think there are some scenarios in which these limited-information heartbeats could cause problems. For example, consider a group containing nodes A, B, and C:
* A is initially the leader.
* Node C loses connectivity to the other nodes and after a timeout declares itself a candidate.
* The partition partially heals; C can talk to A but not B.
* C is elected the new leader; A steps down.
* B sees that A has been up continuously and so it still considers A the leader.

We might be able to solve this on a case-by-case basis (e.g. I think two-round elections would help in the above scenario), but to be safe we should probably add a payload to the coalesced heartbeat messages to indicate which groups the sending node considers itself to be leader of. 

This thread on raft-dev discusses some related issues: https://groups.google.com/forum/?fromgroups#!topic/raft-dev/VgF47vIsezg. The term and log index is occasionally useful (but not, I think, strictly required. Etcd is not currently sending this information with their heartbeats).

tbg committed:
introduce interceptableLocalTransport, refactor simple heartbeat test

currently this still tests only one group, but it sets a good baseline
for basic sanity. Even writing this test took a lot of care; next step
will be unittests and some multi-group tests with fairly explicit (i.e.
unparametrized) setups.

tbg committed:
increment wait for flakey test

tschottdorf commented:
@bdarnell new transport, refactored test inside. not complete yet but that's what i would build on for the rest of the tests.

bdarnell commented on multiraft/heartbeat_test.go:
```diff
@@ -0,0 +1,228 @@
+			// not a heartbeat.
+			return false
+		}
+		cnt[req.Message.From] = from
```
If heartbeatCountMap contained pointers instead of struct values then you could mutate them without going through this read-then-write pattern.

tschottdorf replied:
It's convenient though because I can avoid the usual `v, ok := ...; if !ok { v[key] = ... }` code.

bdarnell commented on multiraft/heartbeat_test.go:
```diff
@@ -0,0 +1,228 @@
+	ec := 2          // extra ticks; due to election triggering.
+	// Ticks of first follower. Hard coded to one in the test below, since any
+	// higher value can lead to new elections which break the test.
+	// Assigning this to a valuable helps to make sense of the formulae below.
```
s/valuable/variable/

tschottdorf replied:
whoops. Done.

bdarnell commented:
LGTM

tbg committed:
add multi-group heartbeat test

tschottdorf commented:
cc @cockroachdb/developers

bdarnell commented:
The new test LGTM. It's complicated, but I don't have any ideas for simplifying it.

tschottdorf commented:
@bdarnell I've added an extra ping and rebased. Do we want to merge it and tackle the safety considerations above separately? I have a feeling that sending around the groups could be tricky as well - thinking of millions of ranges...

tbg committed:
slight improvement to the test as per @bdarnell

bdarnell commented:
Yes, let's go ahead and merge.
